Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/StateSynchronization/CommandService.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView
+{
+    public class CommandService : Singleton<CommandService>,
+        ICommandService
+    {
+        private readonly object lockObject = new object();
+        private Dictionary<string, List<ICommandHandler>> commandHandlers = new Dictionary<string, List<ICommandHandler>>();
+        private List<ICommandHandler> allHandlers = new List<ICommandHandler>();
```

Comment:
>private List<ICommandHandler> allHandlers = new List<ICommandHandler>(); [](start = 8, length = 72)

One thought I had to make this much more efficient for allocations: use an ImmutableList<ICommandHandler>. The way immutable lists work, they create a new reference to a snapshot of the list when you add or remove items without necessarily needing to perform array allocations or copies. Then the CommandHandlers property could safely return 'allHandlers' without having to worry about collection modification during enumeration.

Reply:
Although I haven't used this in a HoloLens application, not sure which .NET Framework versions support ImmutableList.

---
In reply to: [286255732](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4480#discussion_r286255732) [](ancestors = 286255732)

Reply:
System.Collections.Immutable doesn't resolve for me in Unity. For now this probably suffices, but I'll sync with you to better understand whether i can change the .net version to pick up the immutable namespace/classes.